programs/conditional_vault/src/lib.rs

```diff
@@ -369,73 +378,65 @@ pub mod conditional_vault {
-        token::burn(
-            CpiContext::new(
-                accs.token_program.to_account_info(),
-                Burn {
-                    mint: accs.conditional_on_finalize_token_mint.to_account_info(),
-                    from: accs
-                        .user_conditional_on_finalize_token_account
-                        .to_account_info(),
-                    authority: accs.authority.to_account_info(),
-                },
+        // burn from both accounts even though we technically only need to burn from one
+        for (conditional_mint, user_conditional_token_account) in [
+            (
+                &accs.conditional_on_finalize_token_mint,
+                &accs.user_conditional_on_finalize_token_account,
             ),
-            conditional_on_finalize_balance,
-        )?;
-
-        token::burn(
-            CpiContext::new(
-                accs.token_program.to_account_info(),
-                Burn {
-                    mint: accs.conditional_on_revert_token_mint.to_account_info(),
-                    from: accs
-                        .user_conditional_on_revert_token_account
-                        .to_account_info(),
-                    authority: accs.authority.to_account_info(),
-                },
+            (
+                &accs.conditional_on_revert_token_mint,
+                &accs.user_conditional_on_revert_token_account,
             ),
-            conditional_on_revert_balance,
-        )?;
-
-        if vault_status == VaultStatus::Finalized {
-            token::transfer(
-                CpiContext::new_with_signer(
+        ] {
+            token::burn(
+                CpiContext::new(
                     accs.token_program.to_account_info(),
-                    Transfer {
-                        from: accs.vault_underlying_token_account.to_account_info(),
-                        to: accs.user_underlying_token_account.to_account_info(),
-                        authority: accs.vault.to_account_info(),
+                    Burn {
+                        mint: conditional_mint.to_account_info(),
+                        from: user_conditional_token_account.to_account_info(),
+                        authority: accs.authority.to_account_info(),
                     },
-                    signer,
                 ),
-                conditional_on_finalize_balance,
-            )?;
-        } else {
-            token::transfer(
-                CpiContext::new_with_signer(
-                    accs.token_program.to_account_info(),
-                    Transfer {
-                        from: accs.vault_underlying_token_account.to_account_info(),
-                        to: accs.user_underlying_token_account.to_account_info(),
-                        authority: accs.vault.to_account_info(),
-                    },
-                    signer,
-                ),
-                conditional_on_revert_balance,
+                user_conditional_token_account.amount,
```

Comment:
It's probably not necessary but it might have been possible to reuse the pre_conditional_balances here.

Reply:
thought about it, but I'd next an extra param to the for loop